File: BlogApp-RN/src/features/authentication/data/datasources/authDatasourceImplementation.ts
import { AxiosError, AxiosInstance, AxiosResponse } from "axios";
import { Platform } from "react-native";
import {
  AuthException,
  ServerException,
  UnknownException,
} from "../../../../core/errors/exceptions";
import { AuthResponse } from "../models/authResponse";
import { ResponseMessage } from "../models/responseMessage";
import AuthDatasource from "./authDatasource";

class AuthDatasourceImplementation implements AuthDatasource {
  url: string;
  constructor(private httpClient: AxiosInstance) {
    const address = Platform.OS === "ios" ? "localhost" : "10.0.2.2";
    this.url = `http://${address}:3001/api/auth`;
  }

  login: (username: string, password: string) => Promise<AuthResponse> = async (
    username,
    password
  ) => {
    const endpoint = `${this.url}/login`;
    const body: AuthRequest = { username: username, password: password };
    try {
      const response: AxiosResponse<AuthResponse> =
        await this.httpClient.post<AuthResponse>(endpoint, body);
      const authResponse: AuthResponse = response.data;
      return authResponse;
    } catch (exception) {
      if (exception instanceof AxiosError) {
        if (exception.code === "ERR_NETWORK") {
          throw new ServerException("Could not connect to the server", 503);
        }
        throw new AuthException(
          exception.response?.data["message"] as string,
          exception.code
        );
      }
      throw new UnknownException("An Unknown Exception has occurred", 500);
    }
  };

  signUp: (username: string, password: string) => Promise<AuthResponse> =
    async (username, password) => {
      const endpoint = `${this.url}/signUp`;
      const body: AuthRequest = { username, password };
      try {
        const response: AxiosResponse<AuthResponse> =
          await this.httpClient.post<AuthResponse>(endpoint, body);
        return response.data;
      } catch (exception) {
        if (exception instanceof AxiosError) {
          if (exception.code === "ERR_NETWORK") {
            throw new ServerException("Could not connect to the server", 503);
          }
          throw new AuthException(
            exception.response?.data["message"] as string,
            exception.code
          );
        }
        throw new UnknownException("An Unknown Exception has occurred", 500);
      }
    };

  logout: () => Promise<ResponseMessage> = () => {
    const success: ResponseMessage = { message: "success" };
    return Promise.resolve(success);
  };
}


export default AuthDatasourceImplementation;

File: blogapp_flutter/lib/features/authentication/data/datasources/auth_datasource_implementation.dart
import 'dart:io' show Platform;

import 'package:dio/dio.dart';

import '../../../../core/errors/exceptions.dart';
import '../models/auth_request.dart';
import '../models/auth_response.dart';
import '../models/response_message.dart';
import './auth_datasource.dart';

class AuthDatasourceImplementation implements AuthDatasource {
  final Dio httpClient;
  late final String url;

  AuthDatasourceImplementation({required this.httpClient}) {
    final address = Platform.isIOS ? 'localhost' : '10.0.2.2';
    url = 'http://$address:3001/api/auth';
  }

  @override
  Future<AuthResponse> login(String username, String password) async {
    final endpoint = '$url/login';
    final body = AuthRequest(username: username, password: password);

    try {
      final response = await httpClient.post<Map<String, dynamic>>(
        endpoint,
        data: body.toJson(),
      );
      return AuthResponse.fromJson(response.data!);
    } on DioException catch (exception) {
      if (exception.type == DioExceptionType.connectionError) {
        throw ServerException('Could not connect to the server', 503);
      }
      throw AuthException(
        exception.response?.data['message'] as String? ??
            'Authentication failed',
        exception.response?.statusCode ?? 400,
      );
    } catch (exception) {
      throw UnknownException('An Unknown Exception has occurred', 500);
    }
  }

  @override
  Future<AuthResponse> signUp(String username, String password) async {
    final endpoint = '$url/signUp';
    final body = AuthRequest(username: username, password: password);

    try {
      final response = await httpClient.post<Map<String, dynamic>>(
        endpoint,
        data: body.toJson(),
      );
      final responseFromJson = AuthResponse.fromJson(response.data!);
      
      return responseFromJson;
    } on DioException catch (exception) {
      if (exception.type == DioExceptionType.connectionError) {
        throw ServerException('Could not connect to the server', 503);
      }
      throw AuthException(
        exception.response?.data['message'] as String? ??
            'Authentication failed',
        exception.response?.statusCode ?? 400,
      );
    } catch (exception) {
      throw UnknownException('An Unknown Exception has occurred', 500);
    }
  }

  @override
  Future<ResponseMessage> logout() async {
    return ResponseMessage(message: 'success');
  }
}
